mlflow/models/utils.py

```diff
@@ -462,6 +462,52 @@ def _enforce_col_schema(pf_input: PyFuncInput, input_schema: Schema):
     return new_pf_input
 
 
+def _reshape_and_cast_pandas_column_values(name, pd_series, tensor_spec):
+    if tensor_spec.shape[0] != -1 or -1 in tensor_spec.shape[1:]:
+        raise MlflowException(
+            "For pandas dataframe input, the first dimension of shape must be a variable "
+            "dimension and other dimensions must be fixed, but in model signature the shape "
+            f"of {'input ' + name if name else 'the unamed input'} is {tensor_spec.shape}."
+        )
+
+    if np.isscalar(pd_series[0]):
+        if tensor_spec.shape != (-1,):
+            raise MlflowException(
+                f"The input pandas dataframe column '{name}' contains scalar "
+                "values, which requires the shape to be (-1,), but got tensor spec "
+                f"shape of {tensor_spec.shape}.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        return _enforce_tensor_spec(np.array(pd_series, dtype=tensor_spec.type), tensor_spec)
+    elif isinstance(pd_series[0], list) and np.isscalar(pd_series[0][0]):
+        # If the pandas column contains list type values,
+        # in this case, the shape and type information is lost,
+        # so do not enforce the shape and type, instead,
+        # reshape the array value list to the required shape, and cast value type to
+        # required type.
+        err_msg = (
+            f"The values in Input dataframe column '{name}' cannot be converted to expected shape "
+            f"{tensor_spec.shape} and type {tensor_spec.type}"
+        )
+        try:
+            flattened_numpy_arr = np.vstack(pd_series.tolist())
+            reshaped_numpy_arr = flattened_numpy_arr.reshape(tensor_spec.shape).astype(
+                tensor_spec.type
+            )
+        except ValueError:
+            raise MlflowException(err_msg, error_code=INVALID_PARAMETER_VALUE)
+        if len(reshaped_numpy_arr) != len(pd_series):
+            raise MlflowException(err_msg, error_code=INVALID_PARAMETER_VALUE)
```

Comment:
I wonder if we should have a different message here since the above processing stages succeeded but the length of the resolved tensor structure is unmatched to the original input (it might help with debugging if a user is struggling)

Reply:
`len(reshaped_numpy_arr) != len(pd_series)` will only be caused by the value shape mismatch. I think we can reuse the above error message.